### DISCUSSION TITLE
How to model the `spillage` for storage units

### DISCUSSION THREAD

**Discussion by datejada** [TulipaEnergy/TulipaEnergyModel.jl#513] [Ideas]
For hydro storage, it is relevant to have a spillage variable. There are two options to model this:

1. **USING FLOW VARIABLES** Keeping the model general, and if the user wants to create a spillage variable, then he needs to create a new `demand` asset (e.g., ground or sink) and then create a flow from the storage to this asset with an associated cost (e.g., penalty cost of XXX €/unit).
**PROS**: we keep the model general and don't create specific variables to model this feature. In addition, we keep the model general, so there are no changes in the code as it is today for the storage balance (the penalty cost of spillage is in the cost of the flow variable). 
**CONS**: The spillage is a flow; therefore, it is an extra level of abstraction for the users (maybe it is not that bad if we guide the user with a tutorial)
**CHANGES**: we need to change the code to allow the `consumers` balance to be either an `==` or `>=` because in the case of the balance in the spillage, the balance is a greater or equal constraint, not equality.

2. **ADD MORE DATA/PROPERTIES TO THE ASSET**: Create the `spillage` variable explicitly, add it to the storage balance constraints, and add the data to the asset.
**PROS**: The slack variable `spillage` is clearly in the equation because it is explicitly defined as a variable.
**CONS**: To avoid creating the extra variables when we don't want the `spillage` (e.g., in a battery), we need additional input data in the files to identify if it has the option of spillage. In addition, the penalty cost also needs to be defined as a property of the asset.  
**CHANGES**: Add the new variables, the input data, and the outputs from the new input data.

What do you think @gnawin @g-moralesespana @clizbe?

**Comment by g-moralesespana** [comment]
Not clear to me yet what would be better. Anyway we need non-served-energy and may need spillage (oversupply) variables in the Demand/hub assets. If we include them there, perhaps not weird to have them in the storage asset.

**Reply by datejada** [reply]
That's interesting because one can use the same approach for the ENS. In that case, you have a high-cost producer with a flow connected to the demand/hub assets. So, with Tulipa, you don't need the ENS variable explicitly, which has PROS and CONS. I don't know what is best either. I'm sure that we will retreat in any case; we just need the minimum retreat solution 🤣

**Comment by gnawin** [comment]
To me both are acceptable. Although

- Option 2 is not as generic as Option 1. As in if we want to model non-served-energy, we would again need an extra variable which may complicates our input data, unless we use a uniform name such as "slack".
- Option 1 has the advantage of modeling the _physical flows_. For me, it is always conceptually easier if we model something as flows when they are _physical flows_. For spillage, it is real water flows to somewhere. And it has the advantage that if we want to model a chain of hydro units, where spillage is used to feed different units, then we can model the chain as a graph. This resonates with our "Getting rid of the node" discussion, this way we can connect a _spillage_ asset directly with another hydro asset.

**Comment by clizbe** [comment]
From what I understand, option 1 sounds better, because it aligns with oversupply and to avoid adding new "things" to the model unless we absolutely have to. I can imagine there are lots of "features" like this that can pile up and build a mess if we're not careful.
A How-To covering spillage, oversupply, etc should be clear enough!

**Comment by clizbe** [comment]
But also, I'm curious how it works with a >= constraint. When would the optimal be >?

**Reply by datejada** [reply]
You are right; since the spillage will have a high cost, the optimal solution will consider keeping the spillage as low as possible (i.e., `=0`). Only in cases where you can't store more water and need to spill the energy can you do it because the balance in the sink node is `>=` greater or equal constraint. This case could be that you have too much water coming in because it is a wet year (or too much ice melting in the case of the Nordics reservoirs), and even if it is produced at maximum capacity, you can't store more energy, so you need to spill part of it. Also, if you have a reduction in your capacity or your availability due to maintenance (we can model that in Tulipa if we want to 😉 ), you can have situations where you need to spill the water coming to your reservoir that you can't store anymore. So, spillage helps represent these situations and make the problem feasible simultaneously. Does that make sense?

**Comment by clizbe** [comment]
@datejada Can we close this and make an issue? Or is it still in discussion?

**Reply by datejada** [reply]
Yes, this will be covered in #596
